File: apps/desktop/launcher/host-discovery.js
/**
 * LAN host discovery display — keep in sync with packages/shared/src/host-discovery.ts
 */

const STAGESYNC_SERVICE_TITLE_RE = /^stagesync(\s|$)/i;
const DASH_SEMVER_RE = /^(\d+)-(\d+)-(\d+)(?:$|[-.])/;

export function normalizeDiscoveryVersion(raw) {
  if (raw == null) return null;
  let v = String(raw).trim().replace(/^v/i, "");
  if (!v) return null;
  const dash = DASH_SEMVER_RE.exec(v);
  if (dash) {
    v = `${dash[1]}.${dash[2]}.${dash[3]}`;
  }
  return v;
}

export function formatDiscoveryVersionLabel(raw) {
  const v = normalizeDiscoveryVersion(raw);
  return v ? `v${v}` : null;
}

function trimOrNull(value) {
  if (value == null) return null;
  const t = String(value).trim();
  return t || null;
}

function hostFromOrigin(origin) {
  if (!origin) return null;
  const trimmed = String(origin).trim();
  if (!trimmed) return null;
  try {
    const u = new URL(
      trimmed.startsWith("http://") || trimmed.startsWith("https://")
        ? trimmed
        : `http://${trimmed}`,
    );
    const host = u.hostname;
    const port = u.port || (u.protocol === "https:" ? "443" : "80");
    if (!host) return null;
    if (
      (port === "80" && u.protocol === "http:") ||
      (port === "443" && u.protocol === "https:")
    ) {
      return host;
    }
    return `${host}:${port}`;
  } catch {
    return trimmed.replace(/^https?:\/\//i, "") || null;
  }
}

function originDisplay(origin) {
  return hostFromOrigin(origin) ?? String(origin).replace(/^https?:\/\//i, "");
}

function isStagesyncServiceTitle(value) {
  return STAGESYNC_SERVICE_TITLE_RE.test(String(value).trim());
}

export function formatDiscoveryTitle(input) {
  const hostname = trimOrNull(input.hostname);
  if (hostname && !isStagesyncServiceTitle(hostname)) {
    return hostname;
  }

  const fromOrigin = hostFromOrigin(input.origin);
  if (fromOrigin) return fromOrigin;

  const serviceName = trimOrNull(input.serviceName);
  if (serviceName && !isStagesyncServiceTitle(serviceName)) {
    return serviceName;
  }

  return trimOrNull(input.origin) ?? "Host";
}

export function formatDiscoveryMeta(input) {
  const bits = [originDisplay(input.origin)];
  const versionLabel = formatDiscoveryVersionLabel(input.version);
  if (versionLabel) bits.push(versionLabel);
  const project = trimOrNull(input.project);
  if (project && project !== "Brak projektu") bits.push(project);
  return bits.join(" · ");
}

File: apps/performer/android/app/src/main/java/com/stagesync/performer/HostDiscovery.kt
package com.stagesync.performer

/**
 * LAN host discovery display — keep in sync with packages/shared/src/host-discovery.ts
 */
object HostDiscovery {
    private val stagesyncServiceTitleRe = Regex("^stagesync(\\s|$)", RegexOption.IGNORE_CASE)
    private val dashSemverRe = Regex("^(\\d+)-(\\d+)-(\\d+)(?:$|[-.])")

    data class Host(
        val title: String,
        val meta: String,
        val origin: String,
    )

    fun normalizeDiscoveryVersion(raw: String?): String? {
        if (raw.isNullOrBlank()) return null
        var v = raw.trim().removePrefix("v").removePrefix("V")
        if (v.isBlank()) return null
        val dash = dashSemverRe.find(v)
        if (dash != null) {
            v = "${dash.groupValues[1]}.${dash.groupValues[2]}.${dash.groupValues[3]}"
        }
        return v
    }

    fun formatDiscoveryVersionLabel(raw: String?): String? {
        val v = normalizeDiscoveryVersion(raw) ?: return null
        return "v$v"
    }

    fun formatDiscoveryTitle(
        hostname: String?,
        origin: String,
        serviceName: String? = null,
    ): String {
        val host = hostname?.trim()?.takeIf { it.isNotEmpty() }
        if (host != null && !isStagesyncServiceTitle(host)) return host

        hostFromOrigin(origin)?.let { return it }

        val service = serviceName?.trim()?.takeIf { it.isNotEmpty() }
        if (service != null && !isStagesyncServiceTitle(service)) return service

        return origin.trim().ifBlank { "Host" }
    }

    fun formatDiscoveryMeta(
        origin: String,
        version: String? = null,
        project: String? = null,
    ): String {
        val bits = mutableListOf(originDisplay(origin))
        formatDiscoveryVersionLabel(version)?.let { bits.add(it) }
        val proj = project?.trim()?.takeIf { it.isNotEmpty() }
        if (proj != null && proj != "Brak projektu") {
            bits.add(proj)
        }
        return bits.joinToString(" · ")
    }

    fun decodeTxtAttribute(bytes: ByteArray?): String? {
        if (bytes == null || bytes.isEmpty()) return null
        return try {
            String(bytes, Charsets.UTF_8).trim().takeIf { it.isNotEmpty() }
        } catch (_: Exception) {
            null
        }
    }

    fun txtAttribute(attributes: Map<String, ByteArray>?, key: String): String? =
        decodeTxtAttribute(attributes?.get(key))

    private fun isStagesyncServiceTitle(value: String): Boolean =
        stagesyncServiceTitleRe.containsMatchIn(value.trim())

    private fun hostFromOrigin(origin: String?): String? {
        if (origin.isNullOrBlank()) return null
        val trimmed = origin.trim()
        return try {
            val withScheme =
                if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
                    trimmed
                } else {
                    "http://$trimmed"
                }
            val uri = java.net.URI(withScheme)
            val host = uri.host ?: return trimmed.removePrefix("http://").removePrefix("https://")
            val port = if (uri.port > 0) uri.port else 4000
            "$host:$port"
        } catch (_: Exception) {
            trimmed.removePrefix("http://").removePrefix("https://").ifBlank { null }
        }
    }

    private fun originDisplay(origin: String): String =
        hostFromOrigin(origin) ?: origin.removePrefix("http://").removePrefix("https://")
}
